### PR TITLE
sweep dst_index for matmul API (only math core)

### PR DESCRIPTION
**Pull request by skotaracTT**
### Ticket

https://github.com/orgs/tenstorrent/projects/123?pane=issue&itemId=118348523&issue=tenstorrent%7Ctt-llk%7C453
### Problem description

* Sweep dst_index parameter.
* Instead of performing a full sweep of the dst_index for each result's tile_size, we test only the edge cases—specifically, the first and last possible index values in the destination register. This approach avoids inflating the number of test cases to 130K without gaining coverage of new code areas or branches.
### What's changed

Test cases for math_matmul --> 47232
### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### CODE REVIEW

File: tests/python_tests/helpers/param_config.py [hunk 1]
```diff
@@ -271,20 +275,107 @@ def generate_format_aware_matmul_combinations(
...
+    max_tiles = base_tile_limit // capacity_divisor
+    max_index = max_tiles - result_tiles
+
+    if max_index < 0:
+        raise ValueError(
+            f"Too many result tiles ({result_tiles}) for destination capacity ({max_tiles}) with {dest_sync.name}"
+        )
+
+    # Return both edge cases: starting at index 0 and at max allowed index
+    return max_index
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/helpers/param_config.py:354]: This was initially my idea however, I always run `dst_index = 0` explicitly since it's the first edge case, so I don't actually need the function to return it. This way I can handle the 0 case separately and independently of this function, and I don't need to index into a two element list unnecessarily. I will fix the comment on line 353 to reflect this logic.

File: tests/python_tests/helpers/param_config.py [hunk 2]
```diff
@@ -271,20 +275,107 @@ def generate_format_aware_matmul_combinations(
...
 
-                    # Add only the valid ones
-                    combinations.extend(
-                        [
-                            (fmt, dest_acc, dims, stochastic_mode)
-                            for dims in valid_dimensions
-                        ]
-                    )
+                    # Add only the valid ones with dest indices
+                    for dims in valid_dimensions:
```
Comment (fvranicTT) [inline tests/python_tests/helpers/param_config.py:279]: Huh, would this work?
```python
def generate_format_aware_matmul_combinations(
    formats_list: List[FormatConfig],
    dest_acc_modes: List[DestAccumulation],
    all_stochastic_modes: List[StochasticRounding] = None,
    dest_sync_modes: List[DestSync] = [DestSync.Half],
):
    """
    Generate matmul dimension combinations with stochastic rounding support and dst index sweeping.

    Rules:
    1. Format outliers (Float16_b->Float16, Bfp8_b->Float16) MUST use dest_acc=Yes
    2. When dest_acc=Yes: max 4 tiles (32-bit dest register)
    3. When dest_acc=No: max 8 tiles (16-bit dest register)
    4. Exclude cases when: stochastic_rounding enabled for Pack (Pack, All) AND dest_acc == DestAccumulation.No
       AND k_tiles >= 4 AND fmt.input_format in [DataFormat.Float16_b, DataFormat.Float32]
       AND fmt.output_format in [DataFormat.Float16_b, DataFormat.Float32]
        - This specific case models when we have bfloat16 in dst register, when stochastic rounding is enabled the datum lacks mantissa bits
        to absorb the accumulated precision loss from multiple matmul ops acrossmultiple tiles. And due to existence of bug, this specific case sometimes fails,
        for now we will exclude this case.
    5. Dst register indices are swept based on dest_acc and number of tiles in dst setting:
       - Using DestSync.Half mode:
         - max_dst_tiles=8 (16bit) or 4 (32bit)

    Returns: List of (format, dest_acc, dimensions, stochastic_rounding, dst_index) tuples
    """
    if all_stochastic_modes is None:
        all_stochastic_modes = [StochasticRounding.No]

    combinations = []

    # Cache dimension combinations to avoid regenerating them
    dimensions_cache = {}

    # Pre-compute format sets for better performance
    bfloat16_formats = {DataFormat.Float16_b, DataFormat.Float32}
    fpu_stochastic_modes = {StochasticRounding.Fpu, StochasticRounding.All}

    for fmt in formats_list:
        # Pre-compute format-specific values
        base_max_tiles = 4 if is_dest_acc_needed(fmt) else 8
        is_fpu_bfloat16 = (
            fmt.input_format in bfloat16_formats
            and fmt.output_format in bfloat16_formats
        )

        for dest_acc in dest_acc_modes:
            max_tiles = 4 if dest_acc == DestAccumulation.Yes else base_max_tiles

            # Use cached dimensions or generate new ones
            if max_tiles not in dimensions_cache:
                dimensions_cache[max_tiles] = generate_matmul_dimension_combinations(
                    max_tiles
                )
            dimensions_list = dimensions_cache[max_tiles]

            for stochastic_mode in all_stochastic_modes:
                is_fpu_stochastic = stochastic_mode in fpu_stochastic_modes

                # Process dimensions with optional filtering
                should_filter = (
                    is_fpu_stochastic
                    and dest_acc == DestAccumulation.No
                    and is_fpu_bfloat16
                )

                for dims in dimensions_list:
                    inputA_dims, inputB_dims = dims
                    matmul_info = calculate_matmul_dimensions(
                        tuple(inputA_dims), tuple(inputB_dims)
                    )

                    # Skip if filtering is needed and kt_dim >= 4
                    if should_filter and matmul_info["kt_dim"] >= 4:
                        continue

                    result_tiles = matmul_info["output_tile_cnt"]

                    max_dst_idx = calculate_edgecase_dest_indices(
                        dest_acc == DestAccumulation.Yes,
                        result_tiles,
                        dest_sync_modes,
                    )[0]

                    combinations.extend(
                        [
                            (fmt, dest_acc, dims, stochastic_mode, 0),
                            (fmt, dest_acc, dims, stochastic_mode, max_dst_idx),
                        ]
                    )
    return combinations
```

File: tests/python_tests/helpers/param_config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/helpers/param_config.py:339]: I made some changes, the lines are different... Are you referring to this?
```cpp
combinations.extend(
    [
        (fmt, dest_acc, dims, stochastic_mode, 0),
        (
            (fmt, dest_acc, dims, stochastic_mode, max_dst_idx)
            if max_dst_idx != 0
            else None
        ),
    ]
```

File: tests/python_tests/helpers/param_config.py [hunk 3]
```diff
@@ -214,77 +214,151 @@ def generate_format_aware_matmul_combinations(
...
-    Generate matmul dimension combinations with stochastic rounding support.
+    Generate matmul dimension combinations with stochastic rounding support and dst index sweeping.
 
     Rules:
     1. Format outliers (Float16_b->Float16, Bfp8_b->Float16) MUST use dest_acc=Yes
-    2. When dest_acc=Yes: max 4 tiles (32-bit dest register)
-    3. When dest_acc=No: max 8 tiles (16-bit dest register)
+    2. Max tile capacity in dest is 8 (16bit) or 4 (32bit)
+        - When dest_acc=Yes: max 4 tiles (32-bit dest register)
+        - When dest_acc=No: max 8 tiles (16-bit dest register)
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/helpers/param_config.py:226]: This is because this function generates a sweep for both matmul APIs, a lot of parameters overlap so they use the same function for sweeping. Even though `math_matmul` doesn't sweep `DestSync` so we just run the test the standard way utilizing `SyncHalf`. However, `unpack_matmul` does sweep `DestSync` and this gets handled in the pytest file by passing in all the sync modes. 
This could look a little different: we could initialize the dest sync modes in the function based on which API we sweep and instead have a flag for `unpack_matmul` instead of passing in `dest_sync_dimensions`?

File: tests/python_tests/helpers/param_config.py [hunk 4]
```diff
@@ -402,3 +478,13 @@ def calculate_edgecase_dest_indices(
...
     return combinations
+
+
+def get_max_dst_index(
+    edgecase_dest_indices: List[Tuple[DestSync, int]]
+) -> Optional[int]:
+    """
+    Get the max destination index from the list of edgecase destination indices.
+    """
+    max_index = max(index for _, index in edgecase_dest_indices)
```
Comment (fvranicTT) [inline tests/python_tests/helpers/param_config.py:489]:
```suggestion
    max_index = max((index for _, index in edgecase_dest_indices), default=0)
```
If `edgecase_dest_indices` is empty, simple `max(...)` will raise a `ValueError`. This change will save you from troubles.

File: tests/python_tests/helpers/param_config.py [hunk 5]
```diff
@@ -217,78 +217,154 @@ def generate_format_aware_matmul_combinations(
...
+    capacity_divisor = 2 if dest_acc else 1
+    max_tiles = base_tile_limit // capacity_divisor
+    max_index = max_tiles - result_tiles
+
+    if max_index < 0:
+        raise ValueError(
+            f"Too many result tiles ({result_tiles}) for destination capacity ({max_tiles}) with {dest_sync.name}"
+        )
+
+    # Return both edge cases: starting at index 0 and at max allowed index
```
Comment (fvranicTT) [inline tests/python_tests/helpers/param_config.py:341]: Is this comment still valid?

File: tests/python_tests/helpers/param_config.py [hunk 6]
```diff
@@ -217,78 +217,154 @@ def generate_format_aware_matmul_combinations(
...
     if all_stochastic_modes is None:
         all_stochastic_modes = [StochasticRounding.No]
 
     combinations = []
 
+    # Cache dimension combinations to avoid regenerating them
+    dimensions_cache = {}
+
+    # Pre-compute format sets for better performance
+    bfloat16_formats = {DataFormat.Float16_b, DataFormat.Float32}
```
Comment (fvranicTT) [inline tests/python_tests/helpers/param_config.py:250]: Float32 is not a bfloat format. Neither is Float16b.
